### ISSUE TITLE
Issues connecting to Java server

### ISSUE THREAD

**Issue by juanfem** [mdavidsaver/p4p#83]
I am having issues to connect to a Java-based EPICS server using p4p v4. This was not the case with previous 3.5.x versions, so I suspect it might be an PVXS issue instead of a p4p issue.

To reproduce, you can run the following class from the epicsCoreJava examples:
https://github.com/epics-base/exampleJava/blob/master/database/src/org/epics/exampleJava/exampleDatabase/ExampleDatabase.java
```
$ mvn -Dexec.mainClass="org.epics.exampleJava.exampleDatabase.ExampleDatabase" exec:java
```

Then you can test pvget/pvput to the `PVRsoft` PV:
```
$ pvget PVRsoft
PVRsoft <undefined>              0 
$ pvput PVRsoft 1
Old : <undefined>              0 
New : <undefined>              1 
```

while using p4p-4.0.0 I get 
```
python -m p4p.client.cli get PVRsoft 
2022-04-28T14:23:50.219062000 WARN pvxs.client.io Server 192.168.0.182:5075 no supported auth.  try to force 'anonymous'
2022-04-28T14:23:50.262491000 ERR pvxs.tcp.io Server 192.168.0.182:5075 connection closed with socket error 54 : Connection reset by peer
```

with p4p-3.5.3 I can do get and put, although the get gives an error as well:
```
$ python -m p4p.client.cli get PVRsoft 
PVRsoft Thu Jan  1 01:00:00 1970 1.0
Orphaned server message 1 : not a channel request
$ python -m p4p.client.cli put PVRsoft=2
PVRsoft=2 ok
$  pvget PVRsoft                        
PVRsoft <undefined>              2 
```

I think the issue is linked to the Java library not supporting access security, but what puzzles me is that pvget/pvput work just fine.

**Comment by mdavidsaver** [comment]
> I suspect it might be an PVXS issue instead of a p4p issue.

This is almost certainly the case.

> I think the issue is linked to the Java library not supporting access security ...

I have a vague memory that this was never fully implemented for the pvAccessJava server.

> ... what puzzles me is that pvget/pvput work just fine.

Not surprising really.  PVA is a complicated protocol, and I've only tested with the client in pvAccessJava.  The java PVA server has not been widely used.

> To reproduce, you can run the following class from the epicsCoreJava examples

It's going to be a couple of days before I have time to look into this.  If you can easily reproduce the issue, if would likely be enough for you to send me a [packet capture](https://mdavidsaver.github.io/pvxs/details.html#packet-capture) of a successful `pvget` run.

Note that I'd like to include (cleaned) captures in [my collection](https://github.com/mdavidsaver/cashark/tree/master/test) of CA/PVA samples.

**Comment by mdavidsaver** [comment]
My quick attempt to build+run the requested example ends with:

```sh
$ mvn -Dexec.mainClass="org.epics.exampleJava.exampleDatabase.ExampleDatabase" exec:java
... downloads the internet ...
java.lang.ClassNotFoundException: org.epics.exampleJava.exampleDatabase.ExampleDatabase
    at java.net.URLClassLoader.findClass (URLClassLoader.java:476)
    at java.lang.ClassLoader.loadClass (ClassLoader.java:589)
    at java.lang.ClassLoader.loadClass (ClassLoader.java:522)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:246)
    at java.lang.Thread.run (Thread.java:829)
...
$ git ls-files|grep ExampleDatabase
database/src/org/epics/exampleJava/exampleDatabase/ExampleDatabase.java
$ mvn -version
Apache Maven 3.6.3
Maven home: /usr/share/maven
Java version: 11.0.14, vendor: Debian, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.10.0-13-amd64", arch: "amd64", family: "unix"
```

**Comment by mdavidsaver** [comment]
@kasemir @ralphlange Help!  I'm being mystified by maven.  It feels like I am missing some step 0 in the [README](https://github.com/epics-base/exampleJava#building-via-examplejava).  Cloning then running `mvn package` results in an error about a missing `org.epics:epics-util:jar:1.0.0-SNAPSHOT`.

```
... Could not resolve dependencies for project org.epics:epics-examples-database:jar:4.3.0-SNAPSHOT: \
Could not find artifact org.epics:epics-util:jar:1.0.0-SNAPSHOT ...
...
```
Or, is there some other PVA server based on pvAccessJava which is easier to build/run?

**Comment by shroffk** [comment]
Hi,
It looks like the epics-examples-database dependencies have not been updated for quite awhile now.
either we can updated them to point to released binaries or update them to point to the latest snapshot version.

@juanfem what is your preference

**Comment by kasemir** [comment]
To look at this specific issue, I think you really need to get the `ExampleDatabase` from the original Java server implementation to build & run, and Kunal is on the way to get you there.

> is there some other PVA server based on pvAccessJava which is easier to build/run?

Yes, but not easier.

The new Java PVA server/client lib is fundamentally very easy to built.
While the sources are bundled with the rest of phoebus, the lib itself has 
no dependencies and builds with just 'ant':

```
git clone https://github.com/ControlSystemStudio/phoebus.git
cd phoebus/core/pva
ant
# -->  target/core-pva-*.jar
```

It does include a command-line PVA client:
```
java -jar target/core-pva-*.jar 
USAGE: pvaclient info|get|monitor|put [options] <PV name>
...
```

But you're looking for a PVA server. There is a demo server included under src/test/java, but it's not built by the 'ant' script because it has dependencies. I've been running it from within Eclipse. For standalone usage, right now the "easiest" but hacky way would be this:

1) Get the dependencies. While overkill, this is the 2-line recipe:
```
cd phoebus
mvn clean verify  -f dependencies/pom.xml
```

2) Patch the phoebus/core/pva/build.xml to add the 'test' sources
```
-        <javac destdir="${classes}" debug="${debug}">
+        <javac destdir="${classes}" debug="${debug}" classpath="../../dependencies/phoebus-target/target/lib/junit-4.13.1.jar:../../dependencies/phoebus-target/target/lib/hamcrest-core-1.3.jar">
+            <src path="src/test/java"/>
```

3) Run the server demo
```
java -cp target/core-pva-4.6.6.jar org.epics.pva.server.IPv6ServerDemo
# Now look at PV 'demo'
```

**Comment by juanfem** [comment]
Sorry for my late reply, I took some days off...

> @kasemir @shroffk @ralphlange Help! I'm being mystified by maven. It feels like I am missing some step 0 in the [README](https://github.com/epics-base/exampleJava#building-via-examplejava). Cloning then running `mvn package` results in an error about a missing `org.epics:epics-util:jar:1.0.0-SNAPSHOT`.
> 
> ```
> ... Could not resolve dependencies for project org.epics:epics-examples-database:jar:4.3.0-SNAPSHOT: \
> Could not find artifact org.epics:epics-util:jar:1.0.0-SNAPSHOT ...
> ...
> ```
> 
> Or, is there some other PVA server based on pvAccessJava which is easier to build/run?

Here is the full sequence to run the server:
```
$ git clone https://github.com/epics-base/exampleJava.git
$ cd exampleJava/database
$ mvn package
$ mvn -Dexec.mainClass="org.epics.exampleJava.exampleDatabase.ExampleDatabase" exec:java
```
I think that should work...

@kasemir I've just tested that server and it p4p communicates fine with it. Maybe the best is that I port my application to use this new library.

**Comment by kasemir** [comment]
> you can test pvget/pvput to the PVRsoft PV ...
> ..while using p4p-4.0.0 I get .. WARN pvxs.client.io Server .. no supported auth.  try to force 'anonymous'
> ..I've just tested that <new Java> server and it p4p communicates fine with it

This makes some sense.
As we implemented the new C++ (PVXS) and Java libs (phoebus/core/java) we also clarified details of the protocol, https://github.com/epics-base/pvAccessCPP/wiki/protocol, and I think authentication was really added for the first time.
So it's possible that the older Java implementation and the newer C++ implementation don't fully cooperate when it comes to authentication.

**Comment by mdavidsaver** [comment]
With some help I was finally able to build/run the `exampleJava` server.  And it turns out that there is a serious compatibility issue stemming from my mis-reading the PVA  "spec" document, and the existing code.  This manifests when communicating pvAccessJava server, and probably pvAccessCPP on a MSB target (still need to test).

A candidate fix: https://github.com/mdavidsaver/pvxs/pull/27

@juanfem To test this now, you will need to build PVXS as well as P4P from source.

**Comment by mdavidsaver** [comment]
This fix is now released as 4.1.0

@juanfem Could you please confirm that p4p/pvxs now works for you?

**Comment by juanfem** [comment]
@mdavidsaver sorry I did not have time to build and test your candidate fix, but now I've just tested 4.1.0 and the issue is solved. Thanks a lot!
